libsolidity/parsing/Parser.cpp

```diff
@@ -1035,27 +1035,36 @@ ASTPointer<Expression> Parser::parsePrimaryExpression()
 		expression = nodeFactory.createNode<Identifier>(getLiteralAndAdvance());
 		break;
 	case Token::LParen:
+	case Token::LBrack:
 	{
-		// Tuple or parenthesized expression.
-		// Special cases: () is empty tuple type, (x) is not a real tuple, (x,) is one-dimensional tuple
+		// Tuple/parenthesized expression or inline array/bracketed expression.
+		// Special cases: ()/[] is empty tuple/array type, (x)/[] is not a real tuple/array,
+		// (x,) is one-dimensional tuple
 		m_scanner->next();
 		vector<ASTPointer<Expression>> components;
-		if (m_scanner->currentToken() != Token::RParen)
+		Token::Value oppositeToken = (token == Token::LParen ? Token::RParen : Token::RBrack);
+		bool isArray = (token == Token::LBrack ? true : false);
+		if (isArray && (m_scanner->currentToken() == Token::Comma))
```

Comment:
This should be moved inside the while loop - to `components.push_back(ASTPointer<Expression>())` and the error message should be something like `Expected array element`.